tests/HttpClient/Plugin/HybridOauth2Test.php

```diff
@@ -0,0 +1,198 @@
+use GuzzleHttp\Psr7\Response;
+use Http\Client\Exception\TransferException;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * Class HybridOauth2Test.
+ *
+ * @group client
+ * @small
+ */
+class HybridOauth2Test extends TestCase
+{
+    private const API_ENDPOINT = 'http://api.example.com/v1';
+
+    /** @var \Apigee\Edge\Tests\Test\HttpClient\MockHttpClient */
+    protected static $httpClient;
+
+    /** @var \Apigee\Edge\HttpClient\Utility\JournalInterface */
+    private $journal;
+
+    /** @var \Apigee\Edge\ClientInterface */
+    private $client;
+
+    /** @var \Apigee\Edge\HttpClient\Plugin\Authentication\OauthTokenStorageInterface */
+    private $token_storage;
+
+    /**
+     * @inheritdoc
+     */
+    public static function setUpBeforeClass(): void
+    {
+        // Use the Mock HTTP Client for all requests.
+        self::$httpClient = new MockHttpClient();
+        parent::setUpBeforeClass();
+    }
+
+    /**
+     * @inheritdoc
+     */
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->token_storage = new InMemoryOauthTokenStorage();
+        $this->journal = new TestJournal();
+        $this->client = $this->buildClient();
+    }
+
+    /**
+     * @expectedException \Apigee\Edge\Exception\HybridOauth2AuthenticationException
+     * @expectedException \Apigee\Edge\Exception\OauthAccessTokenAuthenticationException
+     */
+    public function testAuthServerError(): void
+    {
+        // Auth server is unavailable.
+        static::$httpClient->addException(new TransferException());
+        $this->client->get('');
+    }
+
+    public function testOauthReAuthenticationAfterExpiredAccessToken(): void
+    {
+        $body = [
+            'access_token' => 'access_token',
+            'expires_in' => 60,
+            'token_type' => 'Bearer',
+        ];
+        // Auth server returns a new access token.
+        static::$httpClient->addResponse(new Response(200, ['Content-Type' => 'application/json'], json_encode((object) $body)));
+
+        // Successful response to an authorised API call from API server.
+        static::$httpClient->addResponse(new Response(200));
+
+        // Another successful response to an authorised API call from API server.
+        static::$httpClient->addResponse(new Response(200));
+
+        $body = [
+            'access_token' => 'access_token',
+            'expires_in' => 60,
+            'token_type' => 'Bearer',
+        ];
+        // Auth server returns a new access token.
+        static::$httpClient->addResponse(new Response(200, ['Content-Type' => 'application/json'], json_encode((object) $body)));
+
+        // Successful response to an authorised API call from API server.
+        static::$httpClient->addResponse(new Response(200));
+
+        $this->client->get('');
+        $this->client->get('');
+
+        // Mark token as expired so the following API call requests a new token.
+        $this->token_storage->markExpired();
```

Comment:
Instead of marking the token expired directly in the storage, mock an API response that would indicate that the token expired, therefore the client has to ask a new one. What is a response for a request on the hybrid server which was authenticated with an expired token? HTTP 401?
https://github.com/apigee/apigee-client-php/blob/2.x/tests/HttpClient/Plugin/OauthTest.php#L94-L112

Reply:
Good suggestion, I've updated it to use the 401 response, which is what the API returns for an expired token.